WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,191 @@
+
+				// Output registration.
+				await RegisterOutputsAsync(outputs).ConfigureAwait(false);
+
+				Transaction? unsignedCoinJoinTransaction = null; // TODO: Get it from somewhere.
+
+				// Send signature.
+				await SignTransactionAsync(aliceClients, unsignedCoinJoinTransaction).ConfigureAwait(false);
+			}
+			catch (Exception ex)
+			{
+			}
+		}
+
+		public async Task StartMixingCoinsAsync()
+		{
+			await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+		}
+
+		private async Task<AliceClient[]> RegisterCoinsAsync()
+		{
+			var aliceArenaClient = new ArenaClient(
+				Round.AmountCredentialIssuerParameters,
+				Round.VsizeCredentialIssuerParameters,
+				AmountCredentialPool,
+				VsizeCredentialPool,
+				ArenaRequestHandler,
+				SecureRandom);
+
+			List<AliceClient> aliceClients = new();
+			try
+			{
+				foreach (var coin in Coins)
+				{
+					// Parallelize or Random delay?
+					aliceClients.Add(await AliceClient.CreateNewAsync(aliceArenaClient, coin, BitcoinSecret, Round.Id, Round.Hash, Round.FeeRate).ConfigureAwait(false));
+				}
+			}
+			catch (Exception)
+			{
+				foreach (var alice in aliceClients)
+				{
+					// Remove already registered Inputs.
+					try
+					{
+						await alice.RemoveInputAsync().ConfigureAwait(false);
+					}
+					catch (Exception)
+					{
+						// Log?
+					}
+				}
+				throw;
+			}
+
+			return aliceClients.ToArray();
+		}
+
+		private async Task ConfirmConnectionsAsync(AliceClient[] aliceClients, CancellationToken stoppingToken)
+		{
+			Task[] confirmTasks = aliceClients.Select(aliceClient => aliceClient.ConfirmConnectionAsync(TimeSpan.FromSeconds(5), stoppingToken)).ToArray();
+
+			await Task.WhenAll(confirmTasks).ConfigureAwait(false);
+
+			var exceptions = confirmTasks.Where(t => t.IsFaulted && t.Exception is { }).Select(t => t.Exception);
+			if (exceptions.Any())
+			{
+				// Error! Try to de-register inputs?
+			}
+		}
+
+		private async Task RegisterOutputsAsync((Money Amount, HdPubKey Pubkey)[] outputs)
+		{
+			ArenaClient bobArenaClient = new(
+				Round.AmountCredentialIssuerParameters,
+				Round.VsizeCredentialIssuerParameters,
+				AmountCredentialPool,
+				VsizeCredentialPool,
+				ArenaRequestHandler,
+				SecureRandom);
+
+			foreach (var output in outputs)
+			{
+				BobClient bobClient = new BobClient(Round.Id, bobArenaClient);
+				await bobClient.RegisterOutputAsync(output.Amount, output.Pubkey.PubKey.WitHash.ScriptPubKey).ConfigureAwait(false);
+				// Random delay?
+			}
+		}
+
+		private async Task<(Money Amount, HdPubKey Pubkey)[]> PlanAmountsAsync(Money[]? inputsOfOthers, CancellationToken stoppingToken)
+		{
+			// Simulate planner.
+			await Task.Delay(0, stoppingToken).ConfigureAwait(false);
+			var amounts = Enumerable.Repeat(Money.Zero, 4).ToArray();
+
+			return amounts.Select(amount => (amount, Keymanager.GenerateNewKey("", KeyState.Locked, true, true))).ToArray(); // Keymanager threadsafe => no!?
```

Comment:
We also need to think about what to do with the generated keys that were used in failed rounds.

IIRC storing these in the wallet was problematic for Wasabi, but I think we can do it more efficiently, and without creating gap limit issues:

- in the normal wallet operations, create change outputs (e.g. from payments and payjoins) by using every 20th address in the internal chain 
- create self spend CJ outputs addresses from the remaining addresses
- renumber these CJ addresses, and encode the ones that were used for failed rounds and never appeared on chain as an interval tree or some other compact representation

This will keep the wallet data reasonably small, if the failure rate is high the representation an interval tree will efficiently encode the dense value range efficiently so regardless of the failure rate the wallet data will scale a logarithmically worse than linearly in the number of successful rounds, instead of linearly in the number of failed rounds. In principle an address can be used a number of times, so long as it's not too many.

It's a little bit crazy, but I think this trick can even be made to work for seed based recovery, without needing to back up the wallet data and while remaining compatible with regular BIP 44 wallets (respecting the gap limit):

- use every 20th address deterministically to ensure other wallets can discover all addresses
- reserve some of the remaining 19 indexes per gap as bits for lossily (conservatively) encoding the interval tree of the exposed-off-chain addresses, for example as an increasing lower bound
- when the wallet data is available the indices can be allocated more efficiently, but upon recovery a supporting wallet can easily know which addresses to avoid ever using again, even if they appear free

Reply:
I think refactoring the Keymanager might be needed but not by us. It can be a huge rabbithole, consider compatibility, tests, and responsibility(one of the highest in the codebase) while we are not going forward with wabisabi. We could delegate the task to pindur or code team. Until then I will try to use the same as we did before. 

https://github.com/molnard/WalletWasabi/blob/01fec42bcde0f60d1b33c1939d0fac9942dfecbb/WalletWasabi/CoinJoin/Client/Clients/CoinJoinClient.cs#L438

Reply:
> It's a little bit crazy, but

It is not crazy and it is almost 100% compatible with what we have right now because Wasabi tries to fill the gaps by default.